### ISSUE TITLE
Expand to ``use-implicit-booleaness-not-len`` to catch `len(iterable) == 0` and `>0`

### ISSUE THREAD

**Issue by DaniBodor** [pylint-dev/pylint#10281]
### Current problem

[C1802](https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/use-implicit-booleaness-not-len.html) currently catches this

Examples 1&2 (currently caught by rule C1802)
```py
fruits = ["orange", "apple"]
vegetables = []

if len(fruits):
    print(fruits)

if not len(vegetables):
    print(vegetables)
```

I come across the following sometimes as well (potentially in botched attempt to fix the the flagged violation above), which I think should be equally flagged as a violation.

Examples 3&4 (not caught by PLC1802)
```py
fruits = ["orange", "apple"]
vegetables = []

if len(fruits) > 0:
    print(fruits)

if len(vegetables) == 0:
    print(vegetables)
```

Examples 5&6 (recommended formulation)
```py
fruits = ["orange", "apple"]
vegetables = []

if fruits:
    print(fruits)

if not vegetables:
    print(vegetables)
```

### Desired solution

Examples 3&4 to be flagged as well

### Additional context

See our previous [discussion](https://github.com/astral-sh/ruff/issues/16798#issuecomment-2729812152) on this issue.

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue, it make sense (especially considering that trying to fix the issue without understanding that empty iterator are truthey often result in code like this)

**Comment by chethanagopinath** [comment]
@Pierre-Sassoulas first timer here, could I pick up this issue?

**Comment by Pierre-Sassoulas** [comment]
Sure, I assigned you to it :)

**Comment by joonhoswe** [comment]
Hi, could i be assigned to this please?

**Comment by Pierre-Sassoulas** [comment]
@chethanagopinath do you still want to work on this issue ?

**Comment by DaniBodor** [comment]
@Pierre-Sassoulas , seems like @chethanagopinath is awol. Maybe re-assign to @joonhoswe ?

**Comment by chethanagopinath** [comment]
@Pierre-Sassoulas apologies for the late reply and delay here, my notifications did not catch the previous tag. Thanks for checking, I had made some progress on a branch locally but couldn't take it all the way, feel free to reassign.

**Comment by mchris86** [comment]
Hi, could I work on this issue?

**Comment by Pierre-Sassoulas** [comment]
Sure, I assigned you to it, let us know if it's a problem 
@joonhoswe

**Comment by Kaos599** [comment]
hey @Pierre-Sassoulas  can i work on this?

**Comment by Pierre-Sassoulas** [comment]
@mchris86 how far along are you, are you still working on it ?

**Comment by Kaos599** [comment]
@Pierre-Sassoulas is it okay if i issue a PR on this?

**Comment by Pierre-Sassoulas** [comment]
Yes go on thank you for working on this :)

**Comment by Kaos599** [comment]
Hey @Pierre-Sassoulas ,
I've submitted PR #10658  extending C1802 to detect len() comparisons with zero (e.g., len(x) == 0, len(x) > 0). 
The implementation adds detection for 6 additional comparison patterns.

can you please review the changes and let me know the feedback, Thanks!

---
If this PR is okay to be merged could you kindly please assign this with `hacktoberfest` label as well?

**Comment by Kaos599** [comment]
@Pierre-Sassoulas  

The C1802 checker was missing len() comparisons with zero that could be simplified using implicit booleaness (e.g., len(seq) == 0 → not seq, len(seq) > 0 → seq). 
I Extended it to catch these patterns while fixing a false positive where compound expressions like return a and len(x) > 0 were incorrectly flagged. Let me know if you'd like any adjustments!

**Comment by Pierre-Sassoulas** [comment]
Let's discuss exclusively in the pull request please.

**Comment by Pierre-Sassoulas** [comment]
This need a decision. As said in https://github.com/pylint-dev/pylint/pull/10658#issuecomment-3419695472. 

We're basically discussing the reversion of 4714a65529a0f03cb77e3938503c131d62fdc9e0 https://github.com/pylint-dev/pylint/issues/2684 / https://github.com/pylint-dev/pylint/pull/2815 here. (Whatever we chose to do we should start with git revert 4714a65529a0f03cb77e3938503c131d62fdc9e0) I should have looked into the history of this and discussed it in more detail before accepting the initial issue. Also sorry for not changing the label here after the discussion in the MR. 

It was also discussed [in the official python discourse](https://discuss.python.org/t/pep8-clarifying-guidance-on-empty-sequence-checks-should-if-len-seq-1-be-replaced-by-if-not-seq/104446). Basically Guido said "make up your mind as pylint maintainer, I don't want to deal with this". With my lecture of PEP8, I think we should revert.

What do you think @jacobtylerwalls @DanielNoord ? 

Marc opinion was:
> Tbh I'm not sure we should emit a message for these. Yes technically they could be rewritten. However, being explicit is often times quite useful.
https://github.com/pylint-dev/pylint/pull/10658#issuecomment-3418391282

**Comment by mvanhorn** [comment]
I've been working on this implementation and wanted to share my perspective.

I think catching `len(seq) == 0` and `len(seq) > 0` is the right call. A few reasons:

1. PEP 8 says "use the fact that empty sequences are false" - `len(seq) == 0` is doing the opposite of that guidance, even if the comparison is explicit.
2. The primer results from my PR show this pattern appears frequently in real codebases (astropy, spack, sympy, etc.), suggesting it's a common anti-pattern worth flagging.
3. As @Pierre-Sassoulas noted, `bool(seq)` is faster than `len(seq) == 0`, so there's a concrete performance benefit.

The strongest counterargument is compound comparisons like `if len(seq) == 0 ... elif len(seq) == 1`. But I'd argue those are a different pattern - comparing to non-zero values is genuinely useful, while comparing to 0 specifically is just a verbose way to check emptiness.

That said, if the decision is to start with `git revert 4714a65` and rebuild from there, I'm happy to rework my PR on top of that.

**Comment by DaniBodor** [comment]
> I think catching `len(seq) == 0` and `len(seq) > 0` is the right call. A few reasons:
> 
>   1. PEP 8 says "use the fact that empty sequences are false" - `len(seq) == 0` is doing the opposite of that guidance, even if the comparison is explicit.
> 
>   2. The primer results from my PR show this pattern appears frequently in real codebases (astropy, spack, sympy, etc.), suggesting it's a common anti-pattern worth flagging.
> 
>   3. As [@Pierre-Sassoulas](https://github.com/Pierre-Sassoulas) noted, `bool(seq)` is faster than `len(seq) == 0`, so there's a concrete performance benefit.
>
> The strongest counterargument is compound comparisons like if len(seq) == 0 ... elif len(seq) == 1. But I'd argue those are a different pattern - comparing to non-zero values is genuinely useful, while comparing to 0 specifically is just a verbose way to check emptiness.

I completely agree with these, and would like to add

4. it is likely (at least anecdotally in my experience) that most cases of `len(seq) == 0` or `len(seq) > 0` are not because the user cares about being explicit, but because they do not notice that there is a more parsimonious way of phrasing it. In case a user truly does want to explicitly compare against 0, they can always decide to suppress the rule (there could even be an argument to have a setting to allow these or not, although I think it would be overkill).

**Comment by jacobtylerwalls** [comment]
Like Marc, I'm also unsure of the value of this.

I added some notes on the "safety" of this kind of fix (it's not always safe) here: https://github.com/astral-sh/ruff/issues/16798#issuecomment-4310006640

Third, there's the issue that we're reverting a prior decision as Pierre mentions here https://github.com/pylint-dev/pylint/issues/10281#issuecomment-4223932806.

**Comment by jacobtylerwalls** [comment]
> I added some notes on the "safety" of this kind of fix (it's not always safe) here: https://github.com/astral-sh/ruff/issues/16798#issuecomment-4310006640

I was wrong, the check is smart enough to only complain about subclasses of `list`, `tuple`, `set`, `str`, etc. So my objection is a little weaker.

**Comment by Pierre-Sassoulas** [comment]
We made a lot of effort to make this check be reasonable (detect ``__bool__``override for pandas and numpy objects and filter based on class inheritance like Jacob said, but also add the 'in typed codebase' warning for all use-impliclt-booleaness like checks -- those are a lot more opinionated, 0 / "" being falsey hurt if you mix string None and float). If we don't have an unreasonable number of false positive for  len(x) >= 1 we're not going to have too much issues  when we entends to zero imo. I'm saying this because the discussion was a lot of 'but it's complicated it's going to have false positives' and we already handle this complexity at the moment. Which was also not the case when the first PR was merged.  

I also feel like using len(x) > 0 instead of len(x) >= 1 is a clumsy way to avoid the warning that is not in the spirit of pep8 nor in the spirit of the pylint check, and doing that make the code look like C. (But it's hard to argue pep8 is authoritative and clear about that given how the discourse discussion went. Guido did say pep8 is clear about this point without elaborating and everyone can hear what they want.)

We could add an option after all ? @DanielNoord what do you think ?

**Comment by DanielNoord** [comment]
Personally I have turned of this check, and this addition would make me turn it off even more 😄 Therefore, I don't really have an opinion.

I don't like how the `if fruits` example would still pass even if you changed `fruits` from `list[...]` to `list[...] | None`. If you add nullability to a variable I'd like the author to have to think about the desired effect on this in other code. The code suggested by this check just silently passes.
For `len(fruits)` I can sort of see that the intent was probably to rely on the `__bool__` attribute for `fruits`, but with `len(fruits) > 0` at least to me it signals "I really expect this always to be something that `len` can handle". To me that reads just fine.

**Comment by Pierre-Sassoulas** [comment]
Looking at the code I realized that "we don't have an unreasonable number of false positive for len(x) >= 1" was based on false assumption, right now we never warn except for raw ``len(x)`` in boolean context.

Seeing 3/4 core maintainer do not like the change, what about adding an option to check comparison against ~= 0 disabled by default ? Or if this is the whole check even len(x) ? In this case if might make more sense to move everything from the refactoring checker to an extension as part of #3512 and add the option enabled by default in the extension.

**Comment by jacobtylerwalls** [comment]
Seems off that we're special-casing zero here.
